Add [Castbar] functionality to GCDTracker

aspiringnobody opened the pull request:
I'm at the point now where I think much more testing is required than I'm able to do.  As long as you don't see any blockers here, I'd like to merge this so we can put it up for testing and get some feedback on what options should stay/go/be moved to some kind of "advanced" tab.  And it needs tested against all the jobs I don't know how to play.  I'm sure much refactoring is needed before this can be completely integrated, but I'm pretty sure all the moving parts are at least there.  I'm going to make one more pass at removing unneeded comments and such, but for now I think I'm ready to move forward.

aspiringnobody commented:
I’ve got one more thing to try to add tonight, and I think that’s it for features.

TODO:
- Show Queuelock on castTime > GCDTotal to visually indicate when you can Queue the next spell.

CaiClone commented:
Hello! Just wanted to drop out to say that I'm keeping an eye on this and it looks great. Really glad to see the option menu simplified further and from my testing the logic has been working great. Once it's ready to review do ping me and I'll get to it!

aspiringnobody commented:
I’m going to have another go at the logic that powers the bar.  I think it will be more maintainable if we use the snapshot at the beginning to define all the possible types of bar (gcd only, cast < gcd, cast == gcd, cast > gcd) and all the combos of settings (queue lock on/off, slidecast on/off, triangles on/off, etc.) and then have a switch block with the required code for each situation.  There will be lots of duplicated code, but I think it makes it a lot easier to reason about what is going on.

I introduced a lot of switching on/off things in a non-intuitive starting state because of code that was above it and I think the extra copypasta is worth it just to be able to see exactly what happens when you’re in one state or another in one clear concise code block.

aspiringnobody commented:
@CaiClone The above is finished now.  I decided to trust the code and just dispense with trying to hide elements that were duplicated but stacked in the same location.  I was having a hard time trusting that the bars would always align perfectly so I had been turning off the sliding elements once they were no longer needed.

In this version I decided to just leave everything on with the understanding that as long as updated in the right order, the pieces should always be in the same place to the pixel.  This lets us get rid of 5 vertex calculations and a bunch of conditional checks trying to orchestrate moving parts.  I think the result is much cleaner, even if we are sometimes drawing the same triangle twice on top of each other.

CaiClone reviewed (changes requested):
Review Sent! General comments are: works well, looks great, can't wait to merge it!

No major blockers in thus review, just comments/requests. I think the biggest concern is the increasing size of the GCDWheel file, which became a 1000 line behemoth, so it may be good to split it up to keep it manageable.

As always, you have already done plenty of development so any part of this review you would prefer me to tackle just close it as is and I'll do it post merge.

CaiClone commented:
Regarding the QueueLock it does come from experimentation as I was not able to find any resource online, so it may be wrong. 0.8 * GCDTime seems to fit everything I have tried, including different Spell Speeds, Leylines and Skills with shorter GCDs such as Mudras, is there any specific case you see failing?

aspiringnobody commented:
The question is essentially does spell/skill speed apply to the 500ms forced ability lock, and to be honest I’m not sure how we might go about testing that. I agree that there doesn’t seem to be much information out there.

I’ll try and see if the XIV Alexander team has a discord, if anyone understands how it really works I bet it would be them.

I’ll let you know if I find anything out.

On Wed, Aug 7, 2024 at 6:26 PM, CaiClone ***@***.***(mailto:On Wed, Aug 7, 2024 at 6:26 PM, CaiClone <<a href=)> wrote:

> Regarding the QueueLock it does come from experimentation as I was not able to find any resource online, so it may be wrong. 0.8 * GCDTime seems to fit everything I have tried, including different Spell Speeds, Leylines and Skills with shorter GCDs such as Mudras, is there any specific case you see failing?
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/CaiClone/GCDTracker/pull/16#issuecomment-2274510972), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AB2ALJ2KKKZ3TXTD6XTLJL3ZQKUIZAVCNFSM6AAAAABL5DP2X2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENZUGUYTAOJXGI).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

aspiringnobody commented on src/Configuration.cs:
```diff
@@ -362,26 +424,115 @@ public class Configuration : IPluginConfiguration
                         BarWidthRatio = size.X;
                         BarHeightRatio = size.Y;
+                        if (ShowAdvanced) {
+                            ImGui.Checkbox("Enable GCDBar Gradient", ref BarHasGradient);
```
Do you have any feedback on what of the modes we should leave in?  I don't mind having the 'extra' options nearly as much now that they're hidden behind the advanced checkbox, but having 3 modes * 2 might be too much?

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -61,18 +71,83 @@ public unsafe class GCDWheel {
+                    return mount != null ? CapitalizeOutput(mount.Singular) : "Unknown Mount";
+
+                default:
+                    //so, we're not going to talk about this, and I'm going to deny ever doing it.
```
I didn't spend any time trying to do this more elegantly because a) we don't need this now that we are getting the game's castbar text directly, and b) I felt like there must be a much better way to do it than to compare the string so I was a garbage implementation anyway.  I left it in since we are still using the case 0 & 1 above to get the queued spell name, and with it all in place we have a quick fallback in case Dalamud ever breaks the ATKStage implementation.

I'll have a look at spiffing it up some.  It's unused at the moment though.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+
+            // Text
+            // reset the queued name when we start to cast.
+            if (castbarProgress <= 0.25f)
```
Maybe it makes sense to move all the cached spell name code there? It gets referenced in DrawGCDBar and DrawCastBar so it probably makes sense to either be with the ability parsing or in its own method called from the ability parser.

aspiringnobody replied:
Changed.  Let me know about the rest.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -61,18 +71,83 @@ public unsafe class GCDWheel {
                     //Store GCD in a variable in order to cache it when it goes back to 0
                     TotalGCD = act->TotalGCD;
                     AddWeaponSkill(act);
+                        queuedAbilityName = GetAbilityName(actionID, DataStore.ClientState.LocalPlayer.CastActionType);
```
I didn't think it was necessary, but I just took a guess where I might put those assignments and never went back to clean it up.  I agree it's not doing anything, but I don't play any classes with weapon skills so I wasn't sure.  Removed.

aspiringnobody commented on src/Configuration.cs:
```diff
@@ -362,26 +424,115 @@ public class Configuration : IPluginConfiguration
+                            ImGui.Checkbox("Enable Slidecast Functionality", ref SlideCastEnabled);
+                            if(SlideCastEnabled) {
+                                ImGui.Indent();
+                                ImGui.Checkbox("Show Slidcast Bar Bacground", ref SlideCastBackground);
```
The whole config needs a pass, specifically better names for the options and on-hover tool-tips created.  Some of the options aren't very intuitive.  I was pretty sure in the beginning that a lot of the options would end up being removed; I'm slightly less sure of that now that there is an advanced toggle to hide most of the junk.

Maybe we can move some of the more niche options to a "very advanced" hidden menu or something?

Or maybe it's fine as it is?

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+            // probably better to move the check for conf.EnableCastText to 
+            // happen before we query the text contents from the game,
+            // but I put it here for now so that we'd be "ready" if someone
+            // was playing with the checkbox while casting (e.g. there'd be stuff
```
I think this is less relevant now that we are using ATKStage.  I doubt the performance hit is anything at all, and the ATKStage approach doesn't require any setup, since the game will always have this data available for us.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+            
+            // in both modes:
+            // draw the background
+            if (!isCastBar)
```
Maybe could even be just 

```
            if (bar.CurrentPos < 0.2f)  
                bgCache = BackgroundColor();  
```

aspiringnobody replied:
Seems to work with just the bar.CurrentPos.  Resolving.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+            }
+
+            //in both modes:
+            //draw the queuelock (if enabled)
```
Actually, that's all handled in the BarDecisionHelper (go.xxxx will only ever be true when it should be drawn).  Judgement call if you want to save the additional checks that happen inside DrawSlideCast and DrawQueueLock by checking if they're needed before we get there.

Judgement call if you want a few more lines of code to save a few conditional checks or not.  It wasn't possible at all until I redid BarDecisionHelper because we sometimes needed bits of the Queuelock even when it was off (after the Slidecast passed where the Queuelock should have been).

But now it is totally possible to check first, if you'd like to do so.

CaiClone replied:
More than saving the checks I was concerned that it was getting a queueLock by having, QueueLockEnabled = False and BarQueueLockWhenIdle = True.

aspiringnobody replied:
Yeah, that was a bug I noticed while working on the mount issue.  I left the check out in BarDecisonHelper ResetBar().  It’s fixed now:

```
Queue_Lock_Start = conf.QueueLockEnabled && conf.BarQueueLockWhenIdle ? 0.8f : 0f;   
Queue_VerticalBar = conf.QueueLockEnabled && conf.BarQueueLockWhenIdle;  
Queue_TopTriangle = conf.QueueLockEnabled && conf.BarQueueLockWhenIdle && conf.ShowQueuelockTriangles;  
```


I put the other check in as well.  Better safe than sorry.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+                TriangleOffset = triangleOffset;
+                IsCastBar = isCastBar;
+                IsShortCast = isShortCast;
+                GCDTotal = DataStore.Action->TotalGCD;
```
I think I fixed this in https://github.com/CaiClone/GCDTracker/commit/4bef91d988dbd35cbfb01003dad7609edaa8ec90

There are some other changes in there from the review as well.  I'll work on it some more tomorrow, but in the interim check and see if this fixes the issue for you.  The mount specifically might always be a bit jerky because the cast time is only 1s.  We are updating in visible chunks per iteration through the loop (as near as I can tell plugins aren't updated in parallel, so we have to wait for all other plugins to have their turn before we get to go again.

This bug would have effected more than just mounts: any interactions, teleports, etc.  Since it was quite a few things I think it best to handle this by creating an entirely new BarState.NonAbilityCast to handle it separately from the other cases.  It has some unique considerations (for example, the Queuelock should always be off).

We could catch this condition in the place you suggest, but I think it might be better to do it in as follows:

```
public void DrawCastBar (PluginUI ui) {  
   DrawBarElements(ui, true, gcdTotal > castTotal, **_gcdTotal < 0.001f**_, castbarProgress * castbarEnd, slidecastStart, slidecastEnd);  
}  
```

This has the downside of requiring an extra hard-coded false in DrawGCDBar:

```
public void DrawGCDBar(PluginUI ui) {  
    DrawBarElements(ui, false, shortCastFinished, false, gcdTime / gcdTotal, gcdTime, gcdTotal);  
}  
```

But I think it makes the most sense to catch it here.  We can then pass this bool onto BarInfo.IsNonAbility and it will be available there in case we need it in another context later.  What do you think?

aspiringnobody replied:
@CaiClone We need to decide if we should show the slidecast on non-ability casts.  DelvUI _does_ and I've emulated that behavior for now (since that's probably what most people are using who would be interested in using our castbar replacement)...

But it doesn't really make sense.  For some non-ability casts, you CAN move in the slidecast window and have the ability fire (like teleports).  In that context slidecast makes sense.  For things like mounts, either the slidecast window is very small, or is inconsistent; slidecasting doesn't work according to the mark we are making. 

In that case, slidecasting doesn't make sense at all.  _But_ DelvUI chooses to do the slidecasting anyway, even for mounts.

We could detect when the cast is a mount and disable it then, but there are probably other cases where slidecasting doesn't apply, and that would turn into a list of exceptions that might need curated with patches.  So I'd lean away from that, unless it is just mounts.

CaiClone replied:
I would be okay with doing it just for mounts. I think that's the big use case and I think it's worth it to disable it for that. I agree that a list of exceptions is the worst case, so unless we can neatly fit everything that doesn't allow slidecast into a recast group or we have a clean way to programmatically detect it just mounts sounds good.

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+            }
+        }
+
+        public class BarInfo {
```
@CaiClone 

This is done, and I also moved PluginUI to UI/* as well as splitting GCDWheel into GCDDisplay and GCDHelper.

Let me know what you think, if it's not to your liking I'll roll it back

aspiringnobody commented on src/GCDWheel.cs:
```diff
@@ -269,51 +341,621 @@ public unsafe class GCDWheel {
+            );
+
+            var go = BarDecisionHelper.Instance;
+                go.Update(bar, conf, isRunning);
```
Done

aspiringnobody commented:
I've done a major rewrite.  Github Desktop ate my lunch while I was trying to make a new branch to not commit this much to the main -- and totally screwed it up.  I think I manually got it back to approximately where it was before Github Desktop messed it up.

As a result, there are probably some indentation errors and the like again.  Probably a good place to pause and have you review again.  I've done pretty much all of the above, just a couple of items I'd like your opinion on.

aspiringnobody commented:
@CaiClone  I added a checkbox in the GCDDisplay tab to enable the world's most basic smoothing algorithm to the progress bar.

See if it helps with the choppiness when summoning your mount.  If it works we can leave it in as an option.  It didn't seem to do much for me, but I figured I'd see what you thought before I deleted it (although it's super minor on my computer anyway).

aspiringnobody commented:
> @CaiClone I added a checkbox in the GCDDisplay tab to enable the world's most basic smoothing algorithm to the progress bar.
> 
> See if it helps with the choppiness when summoning your mount. If it works we can leave it in as an option, off by default. It didn't seem to do much for me, but I figured I'd see what you thought before I deleted it (although it's super minor on my computer anyway).

I did some quantitative testing and decided that there just isn't enough variability in the cycle time to allow the amount of smoothing we can get away with (without significantly altering the bar position) to have any effect.  Even in the worst case scenario, where I made the bar the width of the screen and ran pirme95 to load down my cpu, the iteration to iteration difference was about 40 pixels on a fast cast like my mount.  Enabling the smoothing could, in theory, reduce sudden spikes, but I just wasn't seeing them on my machine in the first place.  I'll see what you think, but IMO it's not worth the extra code.  It just doesn't have any effect 99% of the time.

aspiringnobody commented:
> Regarding the QueueLock it does come from experimentation as I was not able to find any resource online, so it may be wrong. 0.8 * GCDTime seems to fit everything I have tried, including different Spell Speeds, Leylines and Skills with shorter GCDs such as Mudras, is there any specific case you see failing?

Not directly related, but there's a new experimental feature in the advanced tab -- it lets you set an offset for your queue lock to account for ping.  Online sources usually say that this number is 2x your ping (aka round-trip ping time) + 100ms

So for me that's 70ms * 2 + 100 ms = 240ms.  Seems to represent pretty accurately when I can actually hit the button and not clip.  Check it out and let me know if you like it and/or if you think we should finalize it and keep it.  If so, we will need to come up with a better name for it.   The first line represents the last moment you can use an oGCD and not clip.  The second line represents the first moment you can queue an on GCD ability.

CaiClone commented on src/GCDDisplay.cs:
```diff
@@ -0,0 +1,299 @@
+                slidecastEnd = conf.SlideCastFullBar ? 1f : castbarEnd;
+            }
+
+            DrawBarElements(ui, true, gcdTotal > castTotal, gcdTotal < 0.001f, castbarProgress * castbarEnd, slidecastStart, slidecastEnd, castbarEnd);
```
The
` gcdTotal < 0.001f`
fails when casting a nonability while still in GCD from a previous ability Instead we could use a function like this in HelperMethods:
```
        public static bool IsNonAbility() {
            return DataStore.ActionManager->CastActionType
                is not ActionType.Action
                and not ActionType.None;
        }
```
But that doesn't fix the mount thing and I don't think it matters for anything else, so unless you can think of an example I think leaving it like that may be okay.

aspiringnobody commented on src/GCDHelper.cs:
```diff
@@ -0,0 +1,551 @@
+            // Handle caster tax
+            if (!isHardCast && HelperMethods.IsCasting() && DataStore.Action->TotalCastTime - 0.1f >= DataStore.Action->TotalGCD)
+                isHardCast = true;
+            checkABC = !abcBlocker && (idleTimerAccum >= (isHardCast ? (conf.abcDelay + 120) : conf.abcDelay));
```
I had a look today at https://github.com/karashiiro/PingPlugin when I was toying with the idea of automatically setting the ping delay for the QL ping compensation feature I added today.  I decided it was too much to try to implement for just that use, but wondered to myself it if was possible to use data from another plugin (if the user happened to have it enabled/installed).

But from looking at his/her implementation I think it would be a couple hundred lines for us just to grab the current latency.  Didn't seem worth it just for that.

CaiClone commented:
Looks good! Couple of comments on the topics on the extra features:

On the **Queue Lock ping**: I like the idea of having a ping offset, maybe we can also use it for a-b-c delay too if we figure that out, but I think the two lines on the queue lock is a bit confusing as it seems to imply that something has to be done inside that time frame. I would just keep the queue lock on the last position (the queue lock offseted by ping) and remove the first line (last possible weave window). 

I can see how the feature is useful, but I'm a little afraid of overstepping in the "providing dynamically accurate weave windows" might make the game feel worse to play and it's also a bit annoying with things like different animation locks on each skill such as with dragoon, and clip windows that are too small to be noticeable like Fire 3->Blizzard 3.

**Smoothing**: I looked into it a bit more and I think my mount issue was because of mode changes between bars and also the issue of the mount not appearing immediately after cast (which causes the bar to return to GCD mode if rolling) so the smoothing didn't help a lot, I would agree with deleting the smoothing code, and maybe in the future we can find a logical solution, as of now it's not a big issue.

aspiringnobody commented:
> Looks good! Couple of comments on the topics on the extra features:
> 
> On the **Queue Lock ping**: I like the idea of having a ping offset, maybe we can also use it for a-b-c delay too if we figure that out, but I think the two lines on the queue lock is a bit confusing as it seems to imply that something has to be done inside that time frame. I would just keep the queue lock on the last position (the queue lock offseted by ping) and remove the first line (last possible weave window).
> 
> I can see how the feature is useful, but I'm a little afraid of overstepping in the "providing dynamically accurate weave windows" might make the game feel worse to play and it's also a bit annoying with things like different animation locks on each skill such as with dragoon, and clip windows that are too small to be noticeable like Fire 3->Blizzard 3.
> 
> **Smoothing**: I looked into it a bit more and I think my mount issue was because of mode changes between bars and also the issue of the mount not appearing immediately after cast (which causes the bar to return to GCD mode if rolling) so the smoothing didn't help a lot, I would agree with deleting the smoothing code, and maybe in the future we can find a logical solution, as of now it's not a big issue.

I originally did exactly what you suggested (move the queue lock over to the offset position).  What I found was that I didn't realize I was subconsciously using the queue lock line to know when to start mashing the button for the next spell.  I think this applies mostly to casters with short spells (like red mage).  For my bard it doesn't matter as much, but for Red Mage I found it confusing to have the queue lock seemingly show I should be able to enter a new ability but have nothing happen.

That's why I split it apart.  It wasn't what I was set out to do.  I was conflicted about it that's why I left it for you to see before I finalized it.  I was thinking about making another indicator instead, or leaving the line at the 0.8f mark and putting the triangle at the "offset" queue lock.

I wasn't sure if that would be even more confusing since we already had the split indicator.  Maybe I should make a line with an X through it or an octagon/stop sign in the center of the bar to place at the "offset" location?  A completely new indicator to show this is where you "stop" ??

aspiringnobody commented:
QL Ping Compensation is just a slider now -- no other configuration.  As you turn it up, the bar moves left by the specified amount.

aspiringnobody commented:
I had a crash-to-desktop tonight playing with the checkbox to toggle castbar text while casting.  I'll need to investigate why this happened, but since I couldn't reproduce it I've put a check in the config to prevent the state from actually changing if a spell is being cast or if you're in combat.

TODO: Figure out why this caused CTD.
